jm/review-message-number

fivetran-joemarkiewicz opened the pull request:
This PR is intended to be an initial review of the message_enhanced and number_overview models. The comments in this are meant to be responded to and addressed within main. This PR is not meant to be merged.

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -21,9 +21,9 @@
 
-The following table provides a detailed list of all models materialized within this package by default. 
+The following table provides a detailed list of all models materialized within this package by default.  
 
-| **Model**                | **Description**                                                                                                                                |
+| **Model**                |  **Description**                                                                                                                                 |
```
Just wanted to call out here that the models in the table below seem to be left overs from the copied package. I know there is a final clean up task, but may be worth updating this section as you build the models.

fivetran-reneeli replied:
Cleaned! thanks 👍

fivetran-joemarkiewicz replied:
Quick request to lowercase the T in the model names.

fivetran-joemarkiewicz replied:
It looks like this was rather removed entirely. Was that intended?

fivetran-joemarkiewicz replied:
![image](https://github.com/fivetran/dbt_twilio/assets/74217849/8178c796-f740-4228-a007-f76f13329b80)

fivetran-reneeli replied:
my bad-- thanks for catching

fivetran-joemarkiewicz replied:
It looks like the names of the models in the table don't include the double underscores. Other than that, this looks good to go.

fivetran-joemarkiewicz replied:
Thanks for adding the double underscores. However, can you also underscore the T in the names to be consistent with  the model name casing.

fivetran-joemarkiewicz commented on models/twilio_message_enhanced.sql:
```diff
@@ -1,4 +1,4 @@
-with messages as (
+with messages as ( 
```
If we are keeping the `using_twilio_message` variable, then we will need to include a config block on this model to disable it if they are not using messages. 

What is the percent of customers that use messages? I worry that allowing customers to disable the messages component will result in a package with little to no value.

fivetran-reneeli replied:
That's true, it looks like the majority use it, so I actually think removing the config makes the most sense

fivetran-reneeli replied:
removed from source

fivetran-reneeli replied:
Actually, I realize that the data in our internal shared environment doesn't have the `messaging_service` object. Perhaps _should_ leave in the configs?

fivetran-joemarkiewicz replied:
I think we keep the removal of the config for messages, but add one specific for messaging_service as that is just a complementary addition.

fivetran-reneeli replied:
Great, added back the enabled config for messaging_service

fivetran-joemarkiewicz replied:
Thanks! If you add the config however you will also need to add conditionals for wherever you reference the messaging service.

For example this join should only be run if the variable is enabled.

https://github.com/fivetran/dbt_twilio/blob/54b5a438b8df78c95d3aedcdf57315383e00c754/models/twilio__message_enhanced.sql#L61-L62

fivetran-joemarkiewicz replied:
I saw you addressed this!

fivetran-joemarkiewicz commented on models/twilio_message_enhanced.sql:
```diff
@@ -57,12 +57,12 @@ outbound_messages as (
-    union all
-    select * from outbound_messages
-),
+union_messages as ( 
+ 
+    select * from inbound_messages 
+    union all 
+    select * from outbound_messages 
+), 
```
Can we move everything up to the union step to its own intermediate model.

fivetran-reneeli replied:
yeah, thought about making messages into it's own intermediate... moved

fivetran-joemarkiewicz commented on models/twilio_message_enhanced.sql:
```diff
@@ -80,8 +80,8 @@ final as (
         union_messages.message_id,
-        union_messages.messaging_service_id,
-        union_messages.date_sent,
+        union_messages.messaging_service_id, 
```
This is null for all our records. Is that intentional?

Additionally, should we be joining in the messaging_service staging model here to add additional information to the message enhanced model?

fivetran-reneeli replied:
Yeah, it looks like it's null in our message object but that makes sense, the record in our messaging_service doesn't look to be used for Messages (rather, for Conversations) so I assume it's different.

fivetran-reneeli replied:
And good point, yes brought in some relevant fields from that staging model!

fivetran-joemarkiewicz replied:
What do you mean by 

> the record in our messaging_service doesn't look to be used for Messages (rather, for Conversations) so I assume it's different

Does this mean the messaging_service is for a different component and shouldn't be brought into the model? Also if we do end up bringing it into the model, be sure to add a config to turn it off if users don't have the messaging_service component.

fivetran-reneeli replied:
so based on our [ERD](https://docs.google.com/presentation/d/1ZQ3qkzxXX4fFjv1MTLepwsSC6RvEoqVGMufUP52i4Eo/edit#slide=id.g83364f78b7_1_0) it looks like messaging_service can be used for either Message or Conversation, so for our available data it looks like it is for the latter. But we should still keep it in the message model.

fivetran-joemarkiewicz commented on models/twilio_message_enhanced.sql:
```diff
@@ -80,8 +80,8 @@ final as (
-        union_messages.messaging_service_id,
-        union_messages.date_sent,
+        union_messages.messaging_service_id, 
+        union_messages.date_sent, 
```
This is a timestamp rather than a date. Can we rename to be timestamp_sent

fivetran-joemarkiewicz commented on models/twilio_message_enhanced.sql:
```diff
@@ -90,7 +90,7 @@ final as (
         union_messages.body,
         union_messages.num_characters,
-        (union_messages.num_characters- {{ dbt.length("body_no_spaces") }}) - 1 as num_words,
+        (union_messages.num_characters- {{ dbt.length("body_no_spaces") }}) - 1 as num_words, 
```
This is a really interesting approach! I don't have an issue with it, but we should maybe add some logic to ensure if there is not a captured word then we do not result in a negative 1 value (which is the case for all our data). I would expect all the results in our data to show 1 as the massive block of text should be registered as at least 1 word and not -1.

fivetran-reneeli replied:
Ahh, actually I didn't do the math correctly, it should be +1 and not -1. Updated

fivetran-joemarkiewicz commented on models/twilio_message_enhanced.sql:
```diff
@@ -99,13 +99,13 @@ final as (
         union_messages.price_unit,
         union_messages.updated_at,
-        addresses.iso_country
+        addresses.iso_country 
```
This seems to be null for all records. Should we instead try and do the parsing to get the country code from the phone number? Were you able to attempt this approach?

fivetran-joemarkiewicz replied:
Yeah this is a bit tricky. I would prefer we don't use seed data for this as you mentioned this could go stale. Let's brainstorm more about this and circle back.

fivetran-joemarkiewicz commented on models/twilio_message_enhanced.sql:
```diff
@@ -99,13 +99,13 @@ final as (
     left join incoming_phone_number
         on union_messages.phone_number = incoming_phone_number.phone_number
-        and union_messages.phone_number = incoming_phone_number.phone_number
+        and union_messages.phone_number = incoming_phone_number.phone_number 
```
What is the purpose of this second join?

fivetran-reneeli replied:
Update-- I actually removed the address and incoming_phone_number models from this. I believe I misinterpreted the object definitions-- they are records of the _Twilio_ customer, not their customers themselves. I looked at our shared data to confirm

fivetran-joemarkiewicz replied:
It looks like this join was entirely removed. Was that intended?

fivetran-joemarkiewicz replied:
Sounds good!

fivetran-joemarkiewicz commented on models/twilio_number_overview.sql:
```diff
@@ -1,15 +1,15 @@
-        when direction like '%outbound%'
-        then message_id end)
-        as total_outbound_messages,
+    phone_number, 
+    count(case 
+        when direction like '%outbound%' 
+        then message_id end) 
+        as total_outbound_messages, 
```
I imagine we can make this significantly dryer by making this into a for loop. Can you rework to be a for loop.

fivetran-reneeli replied:
Good point, changed to using a for loop

fivetran-joemarkiewicz commented on models/twilio_number_overview.sql:
```diff
@@ -1,15 +1,15 @@
-        when direction like '%outbound%'
-        then message_id end)
-        as total_outbound_messages,
+    phone_number, 
```
Didn't we also want to include country code in this model?

fivetran-reneeli replied:
See earlier note

fivetran-joemarkiewicz replied:
Thanks! We can circle back on this

fivetran-joemarkiewicz commented on models/twilio_number_overview.sql:
```diff
@@ -66,4 +66,4 @@ select
 
 from message_enhanced
-group by phone_number
+group by phone_number 
```
Request to have this be `group by 1`

fivetran-reneeli replied:
updated

fivetran-joemarkiewicz commented on models/twilio_number_overview.sql:
```diff
@@ -1,15 +1,15 @@
-with message_enhanced as (
+with message_enhanced as ( 
```
Reminder that if we add the disable config for the message models then we will need to add a disable config block here.

Again, I am starting to lean to removing that config as now both these models will be removed if the customer doesn't have messaging. I don't see the value in this package if messaging is not used.

fivetran-reneeli replied:
Pending earlier conversation thread

fivetran-joemarkiewicz replied:
Yeah let's plan to remove the config for just messages, but we can add one for messaging_service and account for that accordingly.

fivetran-joemarkiewicz reviewed (commented):
@fivetran-reneeli thanks for working through these models, I left my comments in the notes below. Let me know if you have questions!

fivetran-joemarkiewicz commented:
@fivetran-reneeli just went through and added my responses. Let me know if you have questions.

Also, I didn't catch this previously, but it looks like the model names don't follow our standard format `[app_name]__[model_name]`. Instead these models only have one underscore to separate the app name from the model name. Can you update this as well.

fivetran-joemarkiewicz commented:
Thanks for addressing the comments @fivetran-reneeli! I have a few minor notes, but this looks good to go!

Also be sure to address the model renaming in the tests and what not. See a sample warning here
![image](https://github.com/fivetran/dbt_twilio/assets/74217849/39b0b047-5339-4593-b642-131eca3f89c9)
